enhancements/scheduling/prevent-user-workloads-from-being-scheduled-on-control-plane-nodes.md

```diff
@@ -0,0 +1,504 @@
+
+#### Pre-Workflow Steps
+
+1. Continuous Standardization of Shipped tolerations.
+
+- OpenShift Engineers need to ensure that all control plane and essential workloads ship with the correct tolerations for the NoExecute taints being applied to control plane nodes.
+This is critical for maintaining uninterrupted operations of OpenShift's core services on these nodes.
+This involves getting multiple teams on the same page and that new projects also incorporate the needed tolerations.
+- OpenShift Engineers also need to guarantee that those workloads have the right permissions to tolerate the NoExecute taint, as the ValidatingAdmissionPolicy checks for custom taint/toleration key and verbs.
+- If custom special node groups are being considered for protection, Cluster Administrators need to coordinate with Users/Developers to implement new tolerations for workloads intended for those specific node groups (e.g., GPU-enabled nodes).
+This coordination ensures that the designated workloads are appropriately scheduled on the protected nodes.
+
+#### Workflow Steps
+
+1. Node tainting:
+
+- Cluster Administrator applies NoExecute taints to control plane nodes and any other special node groups deemed necessary to protect.
+This foundational step prevents unauthorized workloads from being scheduled (or directly placed with spec.nodeName field) on these critical nodes.
+
+```
+oc taint nodes ip-XX-XX-XX-XXX.ec2.internal node-role.kubernetes.io/control-plane:NoExecute-
+```
+
+2. Validating Admission Policy provided:
+
+- As part of this proposal a `ValidatingAdmissionPolicy` will be provided as part of the static manifests applied by Kube Scheduler Operator ([staticresourcecontroller](https://github.com/openshift/cluster-kube-scheduler-operator/blob/67304344622e50cbcd95e7a05056294e9d96df99/pkg/operator/starter.go#L108)).
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "control-plane-scheduling-policy"
+spec:
+  failurePolicy: Fail
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   [""]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["pods"]
+  validations:
+    - expression: "object.spec.tolerations.all(toleration, (
+      toleration.effect != 'NoExecute' ||
+      (toleration.effect == 'NoExecute' &&
+      (auhorizer.serviceAccount(object.metadata.namespace, object.spec.serviceAccountName).group('').resource(toleration.key).namespace(object.metadata.namespace).check(toleration.effect).allowed()))))"
+      messageExpression: >
+        "Pod toleration for 'NoExecute' is not authorized for service account '" + object.spec.serviceAccountName + "' in namespace '" + object.metadata.namespace + "'."
+
+```
+
+3. Validating Admission Policy Binding:
+
+- The Cluster Administrator creates a ValidatingAdmissionPolicyBinding to dictate the realms of policy application. Binding shown here affects all namespaces.
```

Comment:
We would make the only allowed change `validationActions` and stomp it otherwise. So they can either enable it as-is or go through the "create their own" process so it's not a free-for-all